common/scala/src/main/scala/whisk/core/connector/Message.scala

```diff
@@ -122,3 +116,84 @@ object PingMessage extends DefaultJsonProtocol {
   def parse(msg: String) = Try(serdes.read(msg.parseJson))
   implicit val serdes = jsonFormat(PingMessage.apply _, "name")
 }
+
+trait EventMessageBody extends Message {}
+
+object EventMessageBody extends DefaultJsonProtocol {
+
+  implicit def format = new JsonFormat[EventMessageBody] {
+    def write(eventMessageBody: EventMessageBody) = eventMessageBody match {
+      case m: Metric     => m.toJson
+      case a: Activation => a.toJson
+    }
+
+    def read(value: JsValue) =
+      if (value.asJsObject.fields.contains("metricName")) {
+        value.convertTo[Metric]
+      } else {
+        value.convertTo[Activation]
+      }
+  }
+}
+
+case class Activation(name: String,
+                      statusCode: Int,
+                      duration: Long,
+                      waitTime: Long,
+                      initTime: Long,
+                      kind: String,
+                      conductor: Boolean,
+                      memory: Int,
+                      cause: Option[ActivationId] = None)
+    extends EventMessageBody {
+  override def serialize = Activation.activationFormat.write(this).compactPrint
+
+  def toJson = Activation.activationFormat.write(this)
+}
+
+object Activation extends DefaultJsonProtocol {
+  def parse(msg: String) = Try(activationFormat.read(msg.parseJson))
+  implicit val activationFormat =
+    jsonFormat(
+      Activation.apply _,
+      "name",
+      "statusCode",
+      "duration",
+      "waitTime",
+      "initTime",
+      "kind",
+      "conductor",
+      "memory",
+      "cause")
+}
+
+case class Metric(metricName: String, metricValue: Long) extends EventMessageBody {
+  override def serialize = Metric.metricFormat.write(this).compactPrint
+  def toJson = Metric.metricFormat.write(this).asJsObject
+}
+
+object Metric extends DefaultJsonProtocol {
+  def parse(msg: String) = Try(metricFormat.read(msg.parseJson))
+  implicit val metricFormat = jsonFormat(Metric.apply _, "metricName", "metricValue")
+}
+
+case class EventMessage(source: String,
+                        body: EventMessageBody,
+                        subject: Subject,
+                        namespace: String,
+                        userId: UUID,
+                        eventType: String,
+                        timestamp: Long = System.currentTimeMillis())
+    extends Message {
+  override def serialize = EventMessage.format.write(this).compactPrint
+}
+
+object EventMessage extends DefaultJsonProtocol {
+  implicit val format =
+    jsonFormat(EventMessage.apply _, "source", "body", "subject", "namespace", "userId", "eventType", "timestamp")
+
+  def parse(msg: String) = format.read(msg.parseJson)
+  def send(producer: MessageProducer, em: EventMessage) = {
+    producer.send("events", em)
+  }
```

Comment:
Does it make sense to create a `UserMetrics` object, which loads the config (move the UserMetricsConfig case class to the same file). The current code conflates schema with actions.